ts/private/ts_project_worker.js

```diff
@@ -0,0 +1,168 @@
+const fs = require('fs')
+const ts = require('typescript')
+const path = require('path')
+const worker = require('@bazel/worker')
+const MNEMONIC = 'TsProject'
+
+// TODO: drop this together with console.time calls
+const console = new globalThis.console.Console(process.stderr, process.stderr)
+
+function noop() {}
+
+function parseArgsFile() {
+    let argsFilePath = process.argv.pop()
+    if (argsFilePath.startsWith('@')) {
+        argsFilePath = argsFilePath.slice(1)
+    }
+    return fs.readFileSync(argsFilePath).toString().split('\n')
+}
+
+/** @type {ts.FormatDiagnosticsHost} */
+const formatDiagnosticHost = {
+    getCanonicalFileName: (path) => path,
+    getCurrentDirectory: ts.sys.getCurrentDirectory,
+    getNewLine: () => ts.sys.newLine,
+}
+
+function printDiagnostics(diagnostics) {
+    worker.log('')
+    worker.log(ts.formatDiagnostics(diagnostics, formatDiagnosticHost))
+}
+
+/** @type {Date} */
+let lastRequestTimestamp
+/** @type {string} */
+let lastRequestArgHash
+/** @type {ts.WatchOfConfigFile<ts.EmitAndSemanticDiagnosticsBuilderProgram>} */
+let program
+/** @type {import("@bazel/worker").Inputs} */
+let previousInputs
+
+/** @type {Map<string, ts.FileWatcherCallback>} */
+let invalidate = new Map()
+
+
+function getProgram(args) {
+    if (lastRequestArgHash !== args.join(' ')) {
+        program = lastRequestArgHash = undefined
+    }
+    if (!program) {
+        const parsedArgs = ts.parseCommandLine(args)
+        const tsconfigPath = args[args.indexOf('--project') + 1]
+
+        const execRoot = path.resolve(process.cwd(), '..', '..', '..')
+
+        /** @type {ts.System} */
+        const sys = {
+            ...ts.sys,
+            write: worker.log,
+            writeOutputIsTTY: false,
+            setTimeout: (callback) => {
+               return setImmediate(callback)
+            },
+            clearTimeout: clearImmediate,
+            watchFile: (filePath, callback) => {
+                const relativeFilePath = path.relative(execRoot, filePath)
+                invalidate.set(relativeFilePath, (kind) => callback(filePath, kind))
+                return {close: () => invalidate.delete(relativeFilePath)}
+            },
+            watchDirectory: (path, callback, recursive, options) => {
+                // TODO: Figure out what to do with directory watchers
+                // tsc seem to be using them for watching the typeroots, node_modules tree, current dir
+            }
+        }
+  
+        const host = ts.createWatchCompilerHost(
+            tsconfigPath,
+            parsedArgs.options,
+            sys,
+            ts.createSemanticDiagnosticsBuilderProgram,
+            noop,
+            noop
+        )
+
+        lastRequestArgHash = args.join(' ')
```

Comment:
should we also add the contents of the tsconfig to the hash so that the program is re-created whenever tsconfig changes? suppose to be correct you'd also need the contents of any transitive tsconfigs that are extended from?

Reply:
Also, this logic has been changed as well in #26